### PR TITLE
Show default values for plugin settings

### PR DESCRIPTION
**Pull request by X9VoiD**
# Changes

- Added `DefaultPropertyValueAttribute`

### CODE REVIEW

**Comment by X9VoiD** [comment]
Since Interpolator have no parameter-less constructor, we can't inspect its default values yet.

File: OpenTabletDriver.Desktop/Reflection/StoredPluginReference.cs [hunk 1]
```diff
@@ -12,6 +12,8 @@ public StoredPluginReference(PluginManager pluginManager, PluginSettingStore sto
 
         private PluginSettingStore PluginSettings { get; }
 
+        public T DirectConstruct<T>() where T : class => base.Construct<T>();
```
Comment (InfinityGhost) [inline OpenTabletDriver.Desktop/Reflection/StoredPluginReference.cs:15]: If we're going to add this function, we might as well add it with the arguments parameters as well.
```cs
public T BaseConstruct<T>() where T : class => base.Construct<T>();
public T BaseConstruct<T>(params object[] args) where T : class => base.Construct<T>(args);
```

File: OpenTabletDriver.Plugin/Tablet/Interpolator/Interpolator.cs [hunk 2]
```diff
@@ -49,11 +47,15 @@ public virtual bool Enabled
...
                 {
+                    Info.Driver.ReportRecieved += HandleReport;
+                    this.reportStopwatch.Start();
                     this.scheduler.Interval = 1000.0f / Hertz;
                     this.scheduler.Start();
                 }
                 else
                 {
+                    Info.Driver.ReportRecieved -= HandleReport;
+                    this.reportStopwatch.Stop();
```
Comment (InfinityGhost) [inline OpenTabletDriver.Plugin/Tablet/Interpolator/Interpolator.cs:58]: Good performance catch

File: OpenTabletDriver.UX/Controls/PluginSettingStoreCollectionEditor.cs [hunk 3]
```diff
@@ -203,38 +213,48 @@ public void Refresh(PluginSettingStore store)
...
+                this.plugin = this.plugins[this.settingStore.Path];
+            }
+
+            private IEnumerable<Control> GetControls()
             {
-                var type = store.GetPluginReference().GetTypeReference<TSource>();
-                return GetControlsForType(store, type);
+                var type = this.settingStore.GetPluginReference().GetTypeReference<TSource>();
+                return GetControlsForType(type);
             }
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/Controls/PluginSettingStoreCollectionEditor.cs:236]: Keep the original function layout, it was much easier to maintain than this.

File: OpenTabletDriver.UX/Controls/PluginSettingStoreCollectionEditor.cs [hunk 4]
```diff
@@ -256,44 +276,7 @@ private Control GetControlForProperty(PluginSettingStore store, PropertyInfo pro
...
-                        // TODO: replace with slider when possible (https://github.com/picoe/Eto/issues/1772)
-                        tb.ToolTip = $"Minimum: {sliderAttr.Min}, Maximum: {sliderAttr.Max}";
-                        tb.PlaceholderText = $"{sliderAttr.DefaultValue}";
-                        if (setting.Value == null)
-                            setting.SetValue(sliderAttr.DefaultValue);
-                    }
-                    return tb;
-                }
-                else if (genericControls.TryGetValue(property.PropertyType, out var getGenericTextBox))
+                if (genericControls.TryGetValue(property.PropertyType, out var getGenericTextBox))
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/Controls/PluginSettingStoreCollectionEditor.cs:279]: RIP attributes, revert this.

**Review by InfinityGhost — CHANGES_REQUESTED** [review]
Maintainability is being lost here, a lot of this needs to be reverted and rethought.

File: OpenTabletDriver.Plugin/Tablet/Interpolator/Interpolator.cs (hunk 2 above)
Comment (AbstractQbit) [reply OpenTabletDriver.Plugin/Tablet/Interpolator/Interpolator.cs:58]: This doesn't affect performance, but does improve `reportMsAvg` value stability

Comment (InfinityGhost) [reply OpenTabletDriver.Plugin/Tablet/Interpolator/Interpolator.cs:58]: Oh right this is the internal stopwatch, not the event timer

File: OpenTabletDriver.Plugin/Tablet/Interpolator/Interpolator.cs [hunk 5]
```diff
@@ -49,11 +47,15 @@ public virtual bool Enabled
...
                 if (value)
                 {
+                    Info.Driver.ReportRecieved += HandleReport;
+                    this.reportStopwatch.Start();
                     this.scheduler.Interval = 1000.0f / Hertz;
                     this.scheduler.Start();
                 }
                 else
                 {
+                    Info.Driver.ReportRecieved -= HandleReport;
```
Comment (AbstractQbit) [inline OpenTabletDriver.Plugin/Tablet/Interpolator/Interpolator.cs:57]: HandleReport can lock itself out at line 89

**Comment by X9VoiD** [comment]
Did a rewrite of the PR to use attributes instead.

File: OpenTabletDriver.Plugin/Attributes/DefaultPropertyValueAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace OpenTabletDriver.Plugin.Attributes
+{
+    public class DefaultPropertyValueAttribute : Attribute
+    {
+        public object Value { get; }
+        public DefaultPropertyValueAttribute(object Value)
+        {
+            this.Value = Value;
+        }
```
Comment (InfinityGhost) [inline OpenTabletDriver.Plugin/Attributes/DefaultPropertyValueAttribute.cs:11]: Fix this formatting
```cs
public DefaultPropertyValueAttribute(object value)
{
    this.Value = value;
}

public object Value { get; }
```

File: OpenTabletDriver.Desktop/Reflection/PluginSetting.cs [hunk 7]
```diff
@@ -35,6 +35,8 @@ private PluginSetting()
         [JsonProperty]
         public JToken Value { set; get; }
 
+        public bool IsValid => Value != null && Value.Type != JTokenType.Null;
```
Comment (InfinityGhost) [inline OpenTabletDriver.Desktop/Reflection/PluginSetting.cs:38]: Would prefer this to be named `HasValue`, `IsValid` implies that something is incorrect where we just want to know if it has a value.

**Review by InfinityGhost — CHANGES_REQUESTED** [review]

**Comment by X9VoiD** [comment]
Requested changes pushed

**Review by InfinityGhost — APPROVED** [review]

**Pull request merged**